pocs/dome/astrohaven_simulator.py

```diff
@@ -0,0 +1,341 @@
+        _drain_queue(self.status_queue)
+
+    def read(self, size=1):
+        """Read size bytes.
+
+        If a timeout is set it may return fewer characters than requested.
+        With no timeout it will block until the requested number of bytes
+        is read.
+
+        Args:
+            size: Number of bytes to read.
+
+        Returns:
+            Bytes read from the port, of type 'bytes'.
+        """
+        if not self.is_open:
+            raise serialutil.portNotOpenError
+
+        # Not checking if the config is OK, so will try to read from a possibly
+        # empty queue if using the wrong baudrate, etc. This is deliberate.
+
+        response = bytearray()
+        timeout_obj = serialutil.Timeout(self.timeout)
+        while True:
+            b = self._read1(timeout_obj)
+            if b:
+                response += b
+                if size is not None and len(response) >= size:
+                    break
+            else:
+                # The timeout expired while in _read1.
+                break
+            if timeout_obj.expired():
+                break
+        return bytes(response)
+
+    @property
+    def out_waiting(self):
+        """The number of bytes in the output buffer."""
+        if not self.is_open:
+            raise serialutil.portNotOpenError
+        return self.command_queue.qsize()
+
+    def reset_output_buffer(self):
+        """Clear output buffer.
+
+        Aborts the current output, discarding all that is in the output buffer.
+        """
+        if not self.is_open:
+            raise serialutil.portNotOpenError
+        _drain_queue(self.command_queue)
+
+    def flush(self):
+        if not self.is_open:
+            raise serialutil.portNotOpenError
+        while not self.command_queue.empty():
+            time.sleep(0.01)
+
+    def write(self, data):
+        """Write the bytes data to the port.
+
+        Args:
+            data: The data to write (bytes or bytearray instance).
+
+        Returns:
+            Number of bytes written.
+
+        Raises:
+            SerialTimeoutException: In case a write timeout is configured for
+                the port and the time is exceeded.
+        """
+        if not self.is_open:
+            raise serialutil.portNotOpenError
+        if not isinstance(data, (bytes, bytearray)):
+            raise ValueError("write takes bytes")
+        data = bytes(data)  # Make sure it can't change.
+        count = 0
+        timeout_obj = serialutil.Timeout(self.write_timeout)
+        for b in data:
+            self.write1(b, timeout_obj)
+            count += 1
+        return count
+
+    # --------------------------------------------------------------------------
+
+    @property
+    def is_config_ok(self):
+        return (self.baudrate == 9600 and self.bytesize == EIGHTBITS and
+                self.parity == PARITY_NONE and not self.rtscts and not self.dsrdtr)
+
+    def _read1(self, timeout_obj):
```

Comment:
Why the `1` in this `_read1` and `_write1`? Is that because they are only dealing with one character? No super important as they are private and this is in the simulator but is none obvious at first glance.